jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildTarballAndLoadDockerStep.java

```diff
@@ -0,0 +1,125 @@
+      BuildConfiguration buildConfiguration,
+      ListeningExecutorService listeningExecutorService,
+      ListenableFuture<List<ListenableFuture<CachedLayer>>> pullBaseImageLayerFuturesFuture,
+      List<ListenableFuture<CachedLayer>> buildApplicationLayerFutures,
+      ListenableFuture<ListenableFuture<Blob>> buildConfigurationFutureFuture) {
+    this.buildConfiguration = buildConfiguration;
+    this.listeningExecutorService = listeningExecutorService;
+    this.pullBaseImageLayerFuturesFuture = pullBaseImageLayerFuturesFuture;
+    this.buildApplicationLayerFutures = buildApplicationLayerFutures;
+    this.buildConfigurationFutureFuture = buildConfigurationFutureFuture;
+  }
+
+  /**
+   * Depends on {@code pullBaseImageLayerFuturesFuture} and {@code buildConfigurationFutureFuture}.
+   */
+  @Override
+  public Void call() throws ExecutionException, InterruptedException {
+    List<ListenableFuture<?>> dependencies = new ArrayList<>();
+    dependencies.addAll(NonBlockingFutures.get(pullBaseImageLayerFuturesFuture));
+    dependencies.addAll(buildApplicationLayerFutures);
+    dependencies.add(NonBlockingFutures.get(buildConfigurationFutureFuture));
+    return Futures.whenAllComplete(dependencies)
+        .call(this::afterPushBaseImageLayerFuturesFuture, listeningExecutorService)
+        .get();
+  }
+
+  /**
+   * Depends on {@code pushAuthorizationFuture}, {@code pushBaseImageLayerFuturesFuture.get()},
+   * {@code pushApplicationLayerFutures}, and (@code
+   * containerConfigurationBlobDescriptorFutureFuture.get()}.
+   */
+  private Void afterPushBaseImageLayerFuturesFuture()
+      throws ExecutionException, InterruptedException, IOException {
+    // Add layers to image tarball
+    TarStreamBuilder builder = new TarStreamBuilder();
+    List<String> layerFiles = new ArrayList<>();
+    for (Future<CachedLayer> cachedLayerFuture :
+        NonBlockingFutures.get(pullBaseImageLayerFuturesFuture)) {
+      Path layerFile = NonBlockingFutures.get(cachedLayerFuture).getContentFile();
+      layerFiles.add(layerFile.getFileName().toString());
+      builder.addEntry(new TarArchiveEntry(layerFile.toFile(), layerFile.getFileName().toString()));
+    }
+    for (Future<CachedLayer> cachedLayerFuture : buildApplicationLayerFutures) {
+      Path layerFile = NonBlockingFutures.get(cachedLayerFuture).getContentFile();
+      layerFiles.add(layerFile.getFileName().toString());
+      builder.addEntry(new TarArchiveEntry(layerFile.toFile(), layerFile.getFileName().toString()));
+    }
+
+    // Add config to tarball
+    Blob containerConfigurationBlob =
+        NonBlockingFutures.get(NonBlockingFutures.get(buildConfigurationFutureFuture));
+    Path tempConfig = Files.createTempFile(null, null);
+    tempConfig.toFile().deleteOnExit();
+    containerConfigurationBlob.writeTo(new BufferedOutputStream(Files.newOutputStream(tempConfig)));
+    builder.addEntry(new TarArchiveEntry(tempConfig.toFile(), "config.json"));
```

Comment:
We might want to add a functionality to `TarStreamBuilder` to add entries that are `Blob`s and not files, so that we don't have to create temporary files here. A TODO for now is good.